### ISSUE TITLE
Start-Process http://localhost:8080 fails to open browser on macOS

### ISSUE THREAD

**Issue by tylerl0706** [PowerShell/PowerShell#5715]
Steps to reproduce
------------------
on macOS:
```powershell
Start-Process http://localhost:8080
```

Expected behavior
-----------------

```none
opens browser to http://localhost:8080
```

Actual behavior
---------------

```none
Start-Process : No such file or directory
At line:1 char:1
+ Start-Process http://localhost:8080
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [Start-Process], Win32Exception
+ FullyQualifiedErrorId : System.ComponentModel.Win32Exception,Microsoft.PowerShell.Commands.StartProcessCommand
```

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.0.0-rc.2
PSEdition                      Core
GitCommitId                    v6.0.0-rc.2
OS                             Darwin 16.7.0 Darwin Kernel Version 16.7.0: Thu Jun 15 17:36:27 PDT 2017; root:xnu-3789.70.16~2/RELEASE_X86_64
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

I tested this on Windows with PSCore 6 and it opened the browser correctly.

FWIW, on macOS, I can run `open http://localhost:8080` and it works.

**Comment by ffeldhaus** [comment]
I just encountered the same problem. As it is quite common to open the Web Browser from a script and `Start-Process $Uri` seems to be a common way, this should be supported.

On the other hand it seems a bit weird that `Start-Process -FilePath $Url` opens a webbrowser as a Url is not a FilePath, but a FilePath can be an Url. Maybe there is a better way or a better Cmdlet can be made available to allow opening a Web Browser.

**Comment by mklement0** [comment]
Indeed: since URLs aren't executables, there is no standardized, cross-platform way in which they can be invoked.

Here's a cross-platform function wrapper that relies on native utilities (none needed on Windows, `open` on macOS, `xdg-open` on Linux distros with freedesktop.org-compliant  desktops):

```powershell
<#
.SYNOPSIS
Opens one or more URIs (URLs) in the default web browser.

.DESCRIPTION
Opens one or more URIs (URLs) in the system's default web browser.
Depending on the platform, this may require 1 call per URI.

.PARAMETER Uri
One or more URIs to open; accepts pipeline input - either as strings, or
via objects that have a .Uri property.

.NOTES
On Linux systems, this function equires the xdg-open utility, 
which comes with distros with freedesktop.org-compliant (X11-based) desktops,
such as Ubuntu and Fedora.

.EXAMPLE
'http://example.org', 'http://unixepoch.com' | Open-Uri
#>
function Open-Uri {
  param(
    [Parameter(Mandatory, ValueFromPipeline, ValueFromPipelineByPropertyName)]
    [string[]] $Uri
  )
  begin {
    $allUris = @()
  }
  process {
    $allUris += $Uri
  }
  end {   
    if ($env:OS -eq 'Windows_NT') { # use Start-Process
      # Note: Start-Process accepts neither pipeline input nor multiple
      #       paths to start.
      $allUris | ForEach-Object { Start-Process -FilePath $_ }
    } elseif ((uname) -eq 'Darwin') { # macOS: use native `open` CLI
      open $allUris
    } else { # Linux: assume that xdg-open is available
      xdg-open $allUris
    }
  }
}
```

**Comment by SteveL-MSFT** [comment]
It might be ok if the file path doesn't exist, we try open/xdg-open, but need to understand any potential side effects.

**Comment by mklement0** [comment]
@SteveL-MSFT: Looking at the name and the parameters of `Start-Process`, my sense is that it's not worth trying to incorporate cross-platform URI-opening functionality - despite the fact that opening URIs with `Start-Process` happens to work on Windows (perhaps inspired by the `cmd.exe`-internal, more abstractly named `start` command).

I can't think of any `Start-Process` parameters (such as `-Wait`, `-Credential`, ...) that make sense in the context of opening URIs across platforms, not least because the underlying native utilities don't have standardized functionality.

In short: I think a separate `Start-Uri` (or `Open-Uri`) cmdlet is called for.

**Comment by SteveL-MSFT** [comment]
If corefx supports `ShellExecute` for non-Windows, we should just get this for "free".

**Comment by mklement0** [comment]
@SteveL-MSFT:

That's a good point.

In fact, it already works in PowerShell Core v6.1.0-preview.3:

```powershell
[System.Diagnostics.Process]::Start([System.Diagnostics.ProcessStartInfo] @{ FileName = 'http://example.org'; UseShellExecute = $True })
```

It looks like CoreFx too uses `open` and `xdg-open` behind the scenes.

One option is to surface this feature via `Start-Process`, similar to how it works in Windows PowerShell, but the question is how, specifically:

* In Windows PowerShell,  the equivalent of  `.UseShellExecute` (the `ShellExecute()` WinAPI function) is used by default behind the scenes, unless certain parameters (`-NoNewWindow`, `RedirectStandard*`, `-Credential`, `-LoadUserProfile`, `-UseNewEnvironment`) are used, which forces use of the `CreateProcess()` WinAPI function instead.
   * This behind-the-scenes logic has a troubling inconsistency with respect to locating executables, as @PetSerAl demonstrates in [this SO answer](https://stackoverflow.com/a/35114544/45375).

* On Unix-like platforms, the logic will have to be different:
  * The decision as to whether or not to set `.UseShellExecute` to `true`  will have to be based solely on the following:
     * If `-ArgumentList` is bound, `.UseShellExecute` cannot be used, because, the way it is implemented it only accepts a single file path / URI.
    * Similarly, combining `.UseShellExecute` with any of the other parameters doesn't seem to make much sense (many of which currently simply crash the session as of PowerShell Core v6.1.0-preview.3).
 
   * With compatible parameters present (read: `-FileName` argument only), `Start-Process` would have to determine whether the argument refers to an executable or not, and only in the latter case set `.UseShellExecute` to `true`.

All in all, existing Windows PowerShell behavior notwithstanding, cramming the features of:

* (a) opening data files with whatever GUI application is registered to handle them

* (b) opening URIs with whatever GUI application  is registered to handle them

into `Start-Process` seems like an awkward fit.
Note that both `open` and `xdg-open` are focused on _just_ (a) and (b), and nothing else.

`Invoke-Item` already does (a) on all platforms, but not (b).

As an aside, `Invoke-Item` is seemingly _only_ implemented for the _filesystem_ provider, where it is _asynchronous_ and should therefore have been named `Start-Item`.

So I wonder if (b) should be folded into `Invoke-Item` / `Start-Item`  for conceptual clarity.

**Comment by mklement0** [comment]
On second thought:

Even the conflation of (a) and (b) above is somewhat arbitrary, so, from a purely conceptual perspective, the following division of labor makes sense to me:

* Use `Start-Process` to launch _executables_, using a variety of options (asynchronous or not, different credentials, hidden, ...)

* Use `Start-Item` (`Invoke-Item`) to open _documents_, asynchronously, using whatever GUI application is registered to handle them by default

* Use `Start-Uri` to open _URIs_, asynchronously, using whatever using whatever GUI application is registered to handle them by default

**Comment by SteveL-MSFT** [comment]
@mklement0 I would agree with you if we didn't already have existing semantics with `Start-Process` and `Invoke-Item`.  Since those already exist with expected behavior, it seems unnecessary to break that contract.  Since corefx supports `UseShellExecute` on non-Windows, we can simply enable it as I see in the code, we explicitly set it to false if we detect we're not on Windows Desktop.

**Comment by mklement0** [comment]
@SteveL-MSFT: 

We don't necessarily have to _break_ the contract:

* Nothing would need to change with respect to `Invoke-Item`, except perhaps to add a `Start-Item` alias.

* On _Windows_, you can enable the same behavior as in Windows PowerShell for a consistent experience on that platform.

* On _Unix_, you cannot blindly default to `.UseShellExecute = true`, and the `-Verb` parameter makes no sense - see below.

On top of that we could introduce a `Start-Uri` cmdlet, as proposed,  and document the clean division of labor proposed above as the way forward.

---

As for your PR / taking advantage of `.UseShellExecute` on Unix-like platforms:

* Enabling the `-Verb` parameter makes no sense, because only the default action (comparable to the `Open` verb on Windows) is supported, and the `.Verb` property is effectively _ignored_ (and `.Verbs` returns a empty string array).

* Defaulting to `.UseShellExecute = true` solely based on the condition that _no arguments_ were passed is ill-advised, because `open` and `xdg-open`, similar to `Invoke-Item`, do not look for `.FileName` values that happen to be _executables_ in the PATH, and you definitely want something like `Start-Process gedit` to continue to work on Linux, for instance, i.e., you definitely want to preserve the ability to call executables in the PATH by filename only.
   * As stated, if you wanted to use `.UseShellExecute` on demand, behind the scenes on Unix, you would have to determine whether `.FileName` refers to an executable in the PATH or not, and use `.UseShellExecute = true` only if _not_.

**Comment by SteveL-MSFT** [comment]
@mklement0 That is a good point on checking if `FileName` is an actual file and use `UseShellExecute` when it isn't to keep a consistent experience.  Agree that `-Verb` doesn't make sense outside of Windows.  I think adding new cmdlets should be a separate issue since this one was originally about opening uri's with `Start-Process`

**Comment by mklement0** [comment]
> Agree that -Verb doesn't make sense outside of Windows, but my PR doesn't make it enabled for non-Windows.

Sorry - I misinterpreted the source code.

> I think adding new cmdlets should be a separate issue since this one was originally about opening uri's with Start-Process

Good point - please see #7201.

**Comment by tylerl0706** [comment]
🎉🎉🎉🎉 Yay!
